File: 1 Algorithmic Toolbox/Week 1 Programming Challenges/2_maximum_pairwise_product/max_pairwise_product.py
# -*- coding: utf-8 -*-
"""
Created on Wed Feb 13 11:25:44 2019

@author: Yu Chen
"""

# python3

# slow but correct algorithm
def max_pairwise_product_naive(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                              numbers[first] * numbers[second])

    return max_product

def max_pairwise_product_fast(numbers):
    n = len(numbers)
    index1 = 0
    for i in range(1, n):
        if (numbers[i] > numbers[index1]):
            index1 = i
    index2 = 0
    for i in range(1, n):
        if ((i != index1) & (numbers[i] > numbers[index2])):
            index2 = i

    return numbers[index1] * numbers[index2]

if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product_naive(input_numbers))
    print(max_pairwise_product_fast(input_numbers))

File: 1 Algorithmic Toolbox/Week 1 Programming Challenges/2_maximum_pairwise_product/max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

// slow but correct algorithm
long long MaxPairwiseProductNaive(const std::vector<long long>& numbers) {
    long long max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

long long MaxPairwiseProductFast(const std::vector<long long>& numbers) {
    int n = numbers.size();
    int index1 = 0;
    int index2 = 0;

    for (int i = 0; i < n; ++i) {
        if (numbers[i] > numbers[index1]){
            index1 = i;
        }
    }
    
    for (int i = 0; i < n; ++i) {
        if ((i != index1) && (numbers[i] > numbers[index2])){
            index2 = i;
        }
    }

    return numbers[index1] * numbers[index2];
}

int main() {
    int n;
    std::cin >> n;
    std::vector<long long> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    long long resultnaive = MaxPairwiseProductNaive(numbers);
    std::cout << resultnaive << "\n";
    long long resultfast = MaxPairwiseProductFast(numbers);
    std::cout << resultfast << "\n";
    return 0;
}
